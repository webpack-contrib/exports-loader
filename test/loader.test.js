import path from 'path';

import {
  compile,
  execute,
  getCompiler,
  getErrors,
  getModuleSource,
  getWarnings,
  readAsset,
} from './helpers';

describe('loader', () => {
  it('should work with string value', async () => {
    const compiler = getCompiler('simple.js', {
      exports: 'Foo',
    });
    const stats = await compile(compiler);

    expect(getModuleSource('./simple.js', stats)).toMatchSnapshot('module');
    expect(
      execute(readAsset('main.bundle.js', compiler, stats))
    ).toMatchSnapshot('result');
    expect(getErrors(stats)).toMatchSnapshot('errors');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
  });

  it('should work with multiple string values', async () => {
    const compiler = getCompiler('simple.js', {
      exports: ['Foo', 'Bar'],
    });
    const stats = await compile(compiler);

    expect(getModuleSource('./simple.js', stats)).toMatchSnapshot('module');
    expect(
      execute(readAsset('main.bundle.js', compiler, stats))
    ).toMatchSnapshot('result');
    expect(getErrors(stats)).toMatchSnapshot('errors');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
  });

  it('should work with inline syntax', async () => {
    // eslint-disable-next-line no-undefined
    const compiler = getCompiler('inline.js', {}, { module: undefined });
    const stats = await compile(compiler);

    expect(getModuleSource('./inline.js', stats)).toMatchSnapshot('module');
    expect(
      execute(readAsset('main.bundle.js', compiler, stats))
    ).toMatchSnapshot('result');
    expect(getErrors(stats)).toMatchSnapshot('errors');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
  });

  it('should work with source maps when the "devtool" option is enabled', async () => {
    const compiler = getCompiler(
      'simple.js',
      {},
      {
        devtool: 'source-map',
        module: {
          rules: [
            {
              test: /\.js$/i,
              rules: [
                {
                  loader: path.resolve(__dirname, '../src'),
                  options: { exports: 'Foo' },
                },
                {
                  loader: 'babel-loader',
                },
              ],
            },
          ],
        },
      }
    );
    const stats = await compile(compiler);

    expect(getModuleSource('./simple.js', stats)).toMatchSnapshot('module');
    expect(
      execute(readAsset('main.bundle.js', compiler, stats))
    ).toMatchSnapshot('result');
    expect(getErrors(stats)).toMatchSnapshot('errors');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
  });

  it('should not work with source maps when the "devtool" options are disabled', async () => {
    const compiler = getCompiler(
      'simple.js',
      {},
      {
        devtool: false,
        module: {
          rules: [
            {
              test: /\.js$/i,
              rules: [
                {
                  loader: path.resolve(__dirname, '../src'),
                  options: { exports: 'Foo' },
                },
                {
                  loader: 'babel-loader',
                },
              ],
            },
          ],
        },
      }
    );
    const stats = await compile(compiler);

    expect(getModuleSource('./simple.js', stats)).toMatchSnapshot('module');
    expect(
      execute(readAsset('main.bundle.js', compiler, stats))
    ).toMatchSnapshot('result');
    expect(getErrors(stats)).toMatchSnapshot('errors');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
  });

  function createSuccessCase(type, exports) {
    it(`should work with the "${type}" module format for ${JSON.stringify(
      exports
    )} export list`, async () => {
      const compiler = getCompiler('simple.js', { type, exports });
      const stats = await compile(compiler);

      expect(getModuleSource('./simple.js', stats)).toMatchSnapshot('module');
      expect(
        execute(readAsset('main.bundle.js', compiler, stats))
      ).toMatchSnapshot('result');
      expect(getErrors(stats)).toMatchSnapshot('errors');
      expect(getWarnings(stats)).toMatchSnapshot('warnings');
    });
  }

  function createFailedCase(type, exports) {
    it(`should work with the "${type}" module format for ${JSON.stringify(
      exports
    )} export list`, async () => {
      const compiler = getCompiler('simple.js', { type, exports });
      const stats = await compile(compiler);

      expect(getErrors(stats)).toMatchSnapshot('errors');
      expect(getWarnings(stats)).toMatchSnapshot('warnings');
    });
  }

  createSuccessCase('commonjs', 'Foo');
  createSuccessCase('commonjs', '[name]');
  createFailedCase('commonjs', 'default');
  createFailedCase('commonjs', 'default Foo');
  createFailedCase('commonjs', 'named Foo');
  createSuccessCase('commonjs', 'single Foo');
  createSuccessCase('commonjs', 'single [name]');
  createSuccessCase('commonjs', 'single single');
  createFailedCase('commonjs', 'single Foo FooA');
  createSuccessCase('commonjs', 'multiple Foo');
  createSuccessCase('commonjs', 'multiple [name]');
  createSuccessCase('commonjs', 'multiple Foo FooA');
  createSuccessCase('commonjs', 'multiple Foo Foo-Bar');
  createSuccessCase('commonjs', 'multiple [name] FooA');
  createSuccessCase('commonjs', 'multiple [name] [name]A');
  createSuccessCase('commonjs', ['Foo', 'Bar']);
  createSuccessCase('commonjs', ['multiple Foo', 'multiple Bar']);
  createSuccessCase('commonjs', ['multiple Foo FooA', 'multiple Bar BarA']);
  createSuccessCase('commonjs', [
    'multiple myVariable.myFunction myFunction',
    'multiple Bar BarA',
  ]);
  createFailedCase('commonjs', 'multiple Foo FooA FooB');
  createFailedCase('commonjs', ['single Foo', 'single Bar']);
  createFailedCase('commonjs', 'unknown Foo');
  createFailedCase('commonjs', '`invalid`');

  createSuccessCase('module', 'Foo');
  createSuccessCase('module', '[name]');
  createFailedCase('module', 'default');
  createFailedCase('module', 'single Foo');
  createFailedCase('module', 'multiple Foo');
  createSuccessCase('module', 'default Foo');
  createSuccessCase('module', 'default [name]');
  createFailedCase('module', 'default default');
  createFailedCase('module', 'default Foo FooA');
  createSuccessCase('module', 'named Foo');
  createSuccessCase('module', 'named [name]');
  createSuccessCase('module', 'named Foo FooA');
  createSuccessCase('module', 'named [name] FooA');
  createSuccessCase('module', 'named [name] [name]A');
  createSuccessCase('module', ['Foo', 'Bar']);
  createSuccessCase('module', ['named Foo', 'named Bar']);
  createSuccessCase('module', ['named Foo FooA', 'named Bar BarA']);
  createSuccessCase('module', ['named Foo default', 'named Bar BarA']);
  createSuccessCase('module', ['default Foo', 'named Bar']);
  createSuccessCase('module', ['default Foo', 'named Bar BarA']);
  createSuccessCase('module', ['default Foo', 'named Bar BarA', 'named Baz']);
  createFailedCase('module', 'named Foo FooA FooB');
  createFailedCase('module', ['default Foo', 'default Bar']);
  createFailedCase('module', 'unknown Foo');
  createFailedCase('module', '`invalid`');
});
