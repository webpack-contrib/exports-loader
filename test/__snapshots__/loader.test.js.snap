// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loader should throw an error without options: errors 1`] = `
Array [
  "ModuleBuildError: Module build failed (from \`replaced original path\`):
ValidationError: Invalid options object. Exports Loader has been initialized using an options object that does not match the API schema.",
]
`;

exports[`loader should throw an error without options: warnings 1`] = `Array []`;

exports[`loader should work for multiple CommonJS exports with "exports": errors 1`] = `Array []`;

exports[`loader should work for multiple CommonJS exports with "exports": module 1`] = `
"const object = {
  key: 'value'
};
const otherObject = {
  otherKey: 'otherValue'
};

exports = { foo: 'bar' };


/*** EXPORTS FROM exports-loader ***/
module.exports = exports = {
  \\"object\\": (object),
  \\"otherObject\\": (otherObject),
};"
`;

exports[`loader should work for multiple CommonJS exports with "exports": result 1`] = `
Object {
  "object": Object {
    "key": "value",
  },
  "otherObject": Object {
    "otherKey": "otherValue",
  },
}
`;

exports[`loader should work for multiple CommonJS exports with "exports": warnings 1`] = `Array []`;

exports[`loader should work for multiple CommonJS exports with "module.exports": errors 1`] = `Array []`;

exports[`loader should work for multiple CommonJS exports with "module.exports": module 1`] = `
"const object = {
  key: 'value'
};
const otherObject = {
  otherKey: 'otherValue'
};

module.exports = { foo: 'bar' };


/*** EXPORTS FROM exports-loader ***/
module.exports = exports = {
  \\"object\\": (object),
  \\"otherObject\\": (otherObject),
};"
`;

exports[`loader should work for multiple CommonJS exports with "module.exports": result 1`] = `
Object {
  "object": Object {
    "key": "value",
  },
  "otherObject": Object {
    "otherKey": "otherValue",
  },
}
`;

exports[`loader should work for multiple CommonJS exports with "module.exports": warnings 1`] = `Array []`;

exports[`loader should work for single CommonJS export with "exports": errors 1`] = `Array []`;

exports[`loader should work for single CommonJS export with "exports": module 1`] = `
"const object = {
  key: 'value'
};
const otherObject = {
  otherKey: 'otherValue'
};

exports = { foo: 'bar' };


/*** EXPORTS FROM exports-loader ***/
module.exports = exports = {
  \\"object\\": (object),
};"
`;

exports[`loader should work for single CommonJS export with "exports": result 1`] = `
Object {
  "object": Object {
    "key": "value",
  },
}
`;

exports[`loader should work for single CommonJS export with "exports": warnings 1`] = `Array []`;

exports[`loader should work for single CommonJS export with "module.exports": errors 1`] = `Array []`;

exports[`loader should work for single CommonJS export with "module.exports": module 1`] = `
"const object = {
  key: 'value'
};
const otherObject = {
  otherKey: 'otherValue'
};

module.exports = { foo: 'bar' };


/*** EXPORTS FROM exports-loader ***/
module.exports = exports = {
  \\"object\\": (object),
};"
`;

exports[`loader should work for single CommonJS export with "module.exports": result 1`] = `
Object {
  "object": Object {
    "key": "value",
  },
}
`;

exports[`loader should work for single CommonJS export with "module.exports": warnings 1`] = `Array []`;

exports[`loader should work with inline syntax: errors 1`] = `Array []`;

exports[`loader should work with inline syntax: module 1`] = `
"module.exports = require('../../src/cjs.js?CV!./exports/cv.js');
"
`;

exports[`loader should work with inline syntax: result 1`] = `
Object {
  "CV": Object {
    "Image": [Function],
    "stackBoxBlurMult": Array [
      1,
      171,
      205,
      293,
      57,
      373,
      79,
      137,
      241,
      27,
      391,
      357,
      41,
      19,
      283,
      265,
    ],
  },
}
`;

exports[`loader should work with inline syntax: warnings 1`] = `Array []`;

exports[`loader should work with source maps: errors 1`] = `Array []`;

exports[`loader should work with source maps: module 1`] = `
"var CV = CV || {};

CV.Image = function(width, height, data){
  this.width = width || 0;
  this.height = height || 0;
  this.data = data || [];
};

CV.stackBoxBlurMult =
  [1, 171, 205, 293, 57, 373, 79, 137, 241, 27, 391, 357, 41, 19, 283, 265];


/*** EXPORTS FROM exports-loader ***/
module.exports = exports = {
  \\"CV\\": (CV),
};"
`;

exports[`loader should work with source maps: result 1`] = `Object {}`;

exports[`loader should work with source maps: warnings 1`] = `Array []`;

exports[`loader should work: errors 1`] = `Array []`;

exports[`loader should work: module 1`] = `
"var CV = CV || {};

CV.Image = function(width, height, data){
  this.width = width || 0;
  this.height = height || 0;
  this.data = data || [];
};

CV.stackBoxBlurMult =
  [1, 171, 205, 293, 57, 373, 79, 137, 241, 27, 391, 357, 41, 19, 283, 265];


/*** EXPORTS FROM exports-loader ***/
module.exports = exports = {
  \\"CV\\": (CV),
};"
`;

exports[`loader should work: result 1`] = `
Object {
  "CV": Object {
    "Image": [Function],
    "stackBoxBlurMult": Array [
      1,
      171,
      205,
      293,
      57,
      373,
      79,
      137,
      241,
      27,
      391,
      357,
      41,
      19,
      283,
      265,
    ],
  },
}
`;

exports[`loader should work: warnings 1`] = `Array []`;
